### PR TITLE
Support for NVIDIA Image Scaling

### PR DESCRIPTION
**Pull request by esullivan-nvidia**
This PR adds support for selecting NVIDIA Image Scaling as the Gamescope upscaler implementation. It can be enabled by passing the -Y or --nis-upscaling option to Gamescope.

### CODE REVIEW

**Comment by emersion** [comment]
Maybe it would make sense to use a submodule or a Meson wrap instead of vendoring the SDK? Or maybe not, it's not that large.

**Comment by Martinligabue** [comment]
why is it removing fsr support?

**Comment by esullivan-nvidia** [comment]
> Maybe it would make sense to use a submodule or a Meson wrap instead of vendoring the SDK? Or maybe not, it's not that large.

I wanted to avoid pulling in the NIS examples directory which would just take up extra space and be unnecessary. Does Meson have a way to exclude some directories when it pulls in a dependency?

**Comment by esullivan-nvidia** [comment]
> why is it removing fsr support?

This change is not removing support for FSR. We are adding support for NIS as an upscaling option. This change will not impact any users who are currently using FSR in Gamescope.

**Comment by eli-schwartz** [comment]
> I wanted to avoid pulling in the NIS examples directory which would just take up extra space and be unnecessary. Does Meson have a way to exclude some directories when it pulls in a dependency?

How would this work from the Meson side? I assume you would download https://github.com/NVIDIAGameWorks/NVIDIAImageScaling/archive/refs/tags/v1.0.2.tar.gz and then extract only the files you need... except that you have already gone and downloaded all 27 mb of that release tarball, which is exactly what you did not want. I don't think there is anything that any application can do about that.

Possible options include:
- not caring about this at all
- re-hosting a stripped tarball, which breaks the association with the upstream release
- asking upstream to publish Github Releases archives of just the SDK
- checking these files into git, but do so in `subprojects/NVIDIAImageScaling/NIS/` instead of `src/shaders/NVIDIAImageScaling/NIS/`, which has the advantage of making it obvious to casual readers that this is imported wholesale from elsewhere rather than code specific to this project, and helps people trivially check the version from the independent `project()`, consult the SBOM, etc.

I would advise at a minimum doing option 4, option 3 would be the best if it can be arranged.

**Review by Joshua-Ashton — APPROVED** [review]
Looks fine to me

**Comment by Joshua-Ashton** [comment]
@DadSchoorse maybe you want to take a look too?

**Comment by Joshua-Ashton** [comment]
(also needs rebasing, sorry :( )

**Review by DadSchoorse — CHANGES_REQUESTED** [review]
Generally it's fine for now, I would like to merge it soon-ish so I can integrate it in my bigger renderer rework.

File: src/rendervulkan.cpp [hunk 1]
```diff
@@ -1287,6 +1308,14 @@ static bool init_device()
...
 	}
 
+	// For NIS if 16 bit shader storage is supported also use the 16 bit coefficients
+	if ( supportsFp16 )
+	{
+		VkFormatProperties rgbaFp16Properties;
+		vkGetPhysicalDeviceFormatProperties(physicalDevice, VK_FORMAT_R16G16B16A16_SFLOAT, &rgbaFp16Properties);
+		g_nisUseFp16Coefficients = ((rgbaFp16Properties.optimalTilingFeatures & VK_FORMAT_FEATURE_SAMPLED_IMAGE_BIT) != 0);
+	}
+
```
[the remark this answers is not in the document]
Comment (esullivan-nvidia) [reply src/rendervulkan.cpp]: Fixed

File: src/rendervulkan.cpp [hunk 2]
```diff
@@ -2171,6 +2249,347 @@ static inline void submit_command_buffer( uint32_t handle, std::vector<std::shar
...
+	samplerInfo.minLod = -1000;
+	samplerInfo.maxLod = 1000;
+	samplerInfo.maxAnisotropy = 1.0f;
+
+	VkResult result = vkCreateSampler(device, &samplerInfo, NULL, &g_output.nisSampler);
+	if(result != VK_SUCCESS)
+	{
+		vk_log.errorf( " failed to create nis sampler " );
+		return false;
+	}
```
[the remark this answers is not in the document]
Comment (esullivan-nvidia) [reply src/rendervulkan.cpp]: Fixed

File: src/rendervulkan.cpp [hunk 3]
```diff
@@ -2171,6 +2249,347 @@ static inline void submit_command_buffer( uint32_t handle, std::vector<std::shar
...
+
+	const void* coef_usm_data = g_nisUseFp16Coefficients ?
+		static_cast<const void*>(coef_usm_fp16) : static_cast<const void*>(coef_usm);
+	uint32_t coef_usm_size = g_nisUseFp16Coefficients ? sizeof(coef_usm_fp16) : sizeof(coef_usm);
+
+	// Create a staging buffer for uploading the data to the GPU
+
+	bufferInfo.size = g_nisUseFp16Coefficients ?	
+		sizeof(coef_scale_fp16) + sizeof(coef_usm_fp16) :
+		sizeof(coef_scale) + sizeof(coef_usm);
```
[the remark this answers is not in the document]
Comment (esullivan-nvidia) [reply src/rendervulkan.cpp]: Fixed

File: src/shaders/descriptor_set.h [hunk 4]
```diff
@@ -11,3 +11,46 @@ layout(binding = 0, rgba8) writeonly uniform image2D dst;
...
+    uint kInputViewportHeight;
+
+    uint kOutputViewportOriginX;
+    uint kOutputViewportOriginY;
+    uint kOutputViewportWidth;
+    uint kOutputViewportHeight;
+
+    float reserved0;
+    float reserved1;
+};
```
[the remark this answers is not in the document]
Comment (esullivan-nvidia) [reply src/shaders/descriptor_set.h:53]: I think we would want to keep this as a uniform buffer to allow Gamescope to keep sharing a single layout between all of its pipelines without adding extra space to the push constant range size.

[the remark this answers is not in the document]
Comment (Joshua-Ashton) [reply src/shaders/descriptor_set.h:53]: Pretty sure we're using like 100 or so bytes of push constant anyway...

[the remark this answers is not in the document]
Comment (DadSchoorse) [reply src/shaders/descriptor_set.h:53]: Yes, our pipeline layout uses the full 128 byte range.

[the remark this answers is not in the document]
Comment (esullivan-nvidia) [reply src/shaders/descriptor_set.h:53]: My bad, I should have double checked that. I can move this over to using push constants.

[the remark this answers is not in the document]
Comment (esullivan-nvidia) [reply src/shaders/descriptor_set.h]: Fixed

**Pull request merged**